### ISSUE TITLE
Device modules not shown on web interface

### ISSUE THREAD

**Issue by glaskoncilla** [netdisco/netdisco#710]
Hi,

we have some number of Cisco ASR9901 and we have strange issue with those devices, device modules are collected during discover (netdisco-do debug including SNMP-info debug said added 551 new chassis modules and there are 551 rows in device_module table related with that particular 9901)  but there is nothing shown under "Module" tab on web interface?

What can cause such kind of issue?

Also, under "Device" tab Model is Products.2658, I've tried to upgrade CISCO-PRODUCTS-MIB.my and CISCO-ENTITY-VENDORTYPE-OID-MIB.my on server with newer version which can for sure resolve Products.2658 using https://github.com/netdisco/netdisco-mibs/wiki/Updating-MIBs as guide.

Basically, I've copy/pasted new version of MIB's over old ones and run mkindex, testload and genxlate scripts, is this enough or I should do something more than this?

Software	Version
App::Netdisco	2.42.10
SNMP::Info	3.68
DB Schema	58
PostgreSQL	9.6.13
Perl	5.28.1

**Comment by glaskoncilla** [comment]
I've managed to update MIB's so Products.2658 model problem is solved but "Module" tab problem still exists.
I've upgraded Netdisco to last version but it didn't help..

**Comment by rc9000** [comment]
Hi! Hmm that is very weird, almost sounds like some string in the module list or it's size crashes the html output.

Could you start the developer tools in Firefox (F12), go to the network tab, click on modules in the device view and see if you get any errors in the request that displays this content?

A successful request should look like this:

![modules](https://user-images.githubusercontent.com/1247031/79796354-b5932680-8355-11ea-823c-dbea245520cf.png)

But I suspect you'll have some 5xx error there. If not, please also go to the Console tab of the developer tools and see if there's anything suspicious there?

**Comment by glaskoncilla** [comment]
Hi,

some string in the module list, now I'm quite sure this is the probem, I've narrow down problem to ASR's 9901 with satellites, 99% of our ASR's has satellites connected and those are the only one with problem, ASR's without satellites are ok.

but this is way over my head of perl to be able to pinpoint exact issue.

output for developer tools;

`<div class="tree">
<ul></ul>
</div>

<script type="text/javascript">
$(document).ready(function() {
$('.tree > ul').attr('role', 'tree').find('ul').attr('role', 'group');
$('.tree').find('li:has(ul)').addClass('parent_li').attr('role', 'treeitem').find(' > span').attr('title', 'Collapse this branch').on('click', function (e) {
        var children = $(this).parent('li.parent_li').find(' > ul > li');
        if (children.is(':visible')) {
     children.hide('fast');
     $(this).attr('title', 'Expand this branch').find(' > i').addClass('icon-plus-sign').removeClass('icon-minus-sign');
        }
        else {
     children.show('fast');
     $(this).attr('title', 'Collapse this branch').find(' > i').addClass('icon-minus-sign').removeClass('icon-plus-sign');
        }
        e.stopPropagation();
    });
});</script>


Error: Error(s) encountered during statement execution: too many SQL variables
Stack trace:
ConnectionData.prototype<._executeStatement/pending<.handleCompletion@resource://gre/modules/Sqlite.jsm:807:25
  Sqlite.jsm:648
Use of getPreventDefault() is deprecated.  Use defaultPrevented instead.  jquery-latest.min.js:4:31310
Use of getAttributeNode() is deprecated. Use getAttribute() instead.  jquery-latest.min.js:5:29041
Empty string passed to getElementById(). jquery-latest.min.js:4:1382
Using //@ to indicate sourceMappingURL pragmas is deprecated. Use //# instead  jquery-latest.min.js:1
GET 
http://netdisco.bnet.lan:5000/javascripts/jquery.min.map [HTTP/1.1 404 Not Found 63ms]
"TabSources.prototype._fetchSourceMap threw an exception: SyntaxError: JSON.parse: unexpected character at line 1 column 1 of the JSON data
Stack: SourceMapConsumer@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/shared/sourcemap/source-map.js:1281:20
TabSources.prototype._fetchSourceMap/fetching<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/server/actors/utils/TabSources.js:469:19
Handler.prototype.process@resource://gre/modules/Promise.jsm -> resource://gre/modules/Promise-backend.js:937:23
this.PromiseWalker.walkerLoop@resource://gre/modules/Promise.jsm -> resource://gre/modules/Promise-backend.js:816:7
this.PromiseWalker.scheduleWalkerLoop/<@resource://gre/modules/Promise.jsm -> resource://gre/modules/Promise-backend.js:750:11
Line: 1281, column: 20"  ThreadSafeDevToolsUtils.js:80
"TabSources.prototype._fetchSourceMap threw an exception: SyntaxError: JSON.parse: unexpected character at line 1 column 1 of the JSON data
Stack: SourceMapConsumer@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/shared/sourcemap/source-map.js:1281:20
TabSources.prototype._fetchSourceMap/fetching<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/server/actors/utils/TabSources.js:469:19
Handler.prototype.process@resource://gre/modules/Promise.jsm -> resource://gre/modules/Promise-backend.js:937:23
this.PromiseWalker.walkerLoop@resource://gre/modules/Promise.jsm -> resource://gre/modules/Promise-backend.js:816:7
this.PromiseWalker.scheduleWalkerLoop/<@resource://gre/modules/Promise.jsm -> resource://gre/modules/Promise-backend.js:750:11
Line: 1281, column: 20"  ThreadSafeDevToolsUtils.js:80
Use of getPreventDefault() is deprecated.  Use defaultPrevented instead.  jquery-latest.min.js:4:31310
"TabSources.prototype._fetchSourceMap threw an exception: SyntaxError: JSON.parse: unexpected character at line 1 column 1 of the JSON data
Stack: SourceMapConsumer@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/shared/sourcemap/source-map.js:1281:20
TabSources.prototype._fetchSourceMap/fetching<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/server/actors/utils/TabSources.js:469:19
Handler.prototype.process@resource://gre/modules/Promise.jsm -> resource://gre/modules/Promise-backend.js:937:23
this.PromiseWalker.walkerLoop@resource://gre/modules/Promise.jsm -> resource://gre/modules/Promise-backend.js:816:7
this.PromiseWalker.scheduleWalkerLoop/<@resource://gre/modules/Promise.jsm -> resource://gre/modules/Promise-backend.js:750:11
Line: 1281, column: 20"  ThreadSafeDevToolsUtils.js:80
Use of getAttributeNode() is deprecated. Use getAttribute() instead.  jquery-latest.min.js:5:29041` 

we can see there "too many SQL variables error"?

also, I can provide complete device_module output for one of the problematic devices if it would be useful for somebody with more perl ninjitzu skills?

**Comment by rc9000** [comment]
Huh this is not even Perl, apparently the JS library in the web client is using local storage (Sqlite.jsm) to render this stuff and that gets tripped up. 

So there's probably stuff in there that somehow ends up as SQL injection from Entity-MIB. The most obvious candidates are patterns like "?", "?NNN", ":AAA", "$AAA", or "@AAA". If you see any of those in there, you can try to delete these lines and see if something is shown afterwards (delete from device_module where 'ip' = xxx and index = yyy).

If it's not obvious, could you try to post a file that we can import. I think the easiest way would be:

 * in psql: `create table export_temp as select * from device_module where ip = 'x.x.x.x'; `
 * then use pg_dump to make an sql file: `pg_dump -d netdisco -t export_temp -U netdisco > export_temp.sql` and attach the file here. 
 * you might want to obfuscate serial numbers and public IP ranges

**Comment by glaskoncilla** [comment]
I didn't find anything suspicious, I've uplodaded dump of one of the problematic devices.



[export_temp.txt](https://github.com/netdisco/netdisco/files/4509112/export_temp.txt)

**Comment by rc9000** [comment]
The issue is that the device does not return a valid tree, there is no root element without a parent. Index 1 is usually a root entity, but here it looks like this:

	{
		"index" : 1,
		"description" : "ASR 9901 Route processor",
		"parent" : 8384602
	},
	{
		"index" : 8384602,
		"description" : "ASR9K Route Processor Slot",
		"parent" : 8384513
	}
	{
		"index" : 8384513,
		"description" : "ASR 9901 Chassis",
		"parent" : 57067179
	},

And 57067179 is nowhere to be found. It's probably an Entity-MIB index in another device, similar things can also happen switch stacks. 

You can get it to work by nulling the two parents that don't exist in the device:

    update device_module set parent = null where ip = '10.10.10.25' and index in (
	SELECT me.index FROM device_module me 
	left join device_module p on me.parent = p.index 
	WHERE me.ip = '10.10.10.25' and p.index is null );

Then it looks as expected:

![scr 2020-04-22 at 00 37 15](https://user-images.githubusercontent.com/1247031/79921706-1f353280-8433-11ea-872b-bfb7ddaa06f9.png)

If this is a common issue, we could probably do something like the above query at discovery time, to make sure no invalid trees end up in device_module. 

Maybe you can also check if you find the device that has the index 57067179 somewhere, and if you discover it you might get a complete module view there.

**Comment by rc9000** [comment]
If you want to implement a private fix, the above query is a bit cumbersome. You can fix the whole table at once by just running this once after discovery:

    update device_module dm set parent = (select index from device_module p where dm.ip = p.ip and dm.parent = p.index);

This is not very efficient and obviously quite untested, but if something breaks you can always just rerun discovery.

**Comment by glaskoncilla** [comment]
I can confirm this is the issue, when you connect the satellite, tree root should become Multi-chassis virtual stack / cevStackIosXrMultichassis, ASR's 9001 with satellites are handling that as expected, and this also explains why ASR's 9901 without satellites don't have this problem.

I will try to investigate this further but I quess some sanatisation code against buggy SNMP implementations of different type of stacks would be enhancement.. 

Thank you for your help, I owe you a bier.

**Comment by rc9000** [comment]
Yeah I also ran some queries against our collection of Cisco gear, we have similar cases with FEX transceivers on Nexus 7700 that have invalid parents. Not as bad since there is still a valid root, they are just invisible. Applying the SQL from above at least makes them show up as disconnected entities at the top level.

![netdisco modules null](https://user-images.githubusercontent.com/1247031/79966571-e8d8d100-848d-11ea-9b3b-03a884162bc7.png)

I'll try to incorporate this workaround, might take a moment though.

**Comment by rc9000** [comment]
Fixed with validation of module parents in Worker/Plugin/Discover/Entities.pm (dcb45e4), seems to catch  all the cases  mentioned here. 

Sorry for messing up the repository history a bit, I committed the wrong file at first - Web/Plugin/Device/Modules.pm does not need any change at all.

**Comment by ollyg** [comment]
awesome, thank you @rc9000
